src/coreclr/tools/Common/Compiler/DependencyAnalysis/ShadowGenericMethodNode.cs

```diff
@@ -42,12 +42,11 @@ public void AppendMangledName(NameMangler nameMangler, Utf8StringBuilder sb)
         public override bool StaticDependenciesAreComputed
             => CanonicalMethodNode.StaticDependenciesAreComputed;
 
-        public ShadowConcreteMethodNode(MethodDesc method, IMethodNode canonicalMethod)
+        public ShadowGenericMethodNode(MethodDesc method, IMethodNode canonicalMethod)
         {
-            Debug.Assert(!method.IsSharedByGenericInstantiations);
```

Comment:
These asserts were useful and prevented nonsensical graphs.

I think we should split this class into multiple classes - one that tracks fully instantiated things (can keep calling it `ShadowConcreteMethod`) - this one can still assert all these things. And add one that is not fully specialized. There may be a common base type if it's useful.

Reply:
Split into ShadowConcreteMethodNode and ShadowNonConcreteMethodNode (with shared base ShadowMethodNode)